src/types/closure.md

```diff
@@ -42,48 +50,153 @@ so that the call to `f` works as if it were:
 
 <!-- ignore: continuation of above -->
 ```rust,ignore
-f(Closure{s: s, t: &t});
+f(Closure{ left_top: rect.left_top, right_bottom_x: rect.left_top.x });
 ```
 
 ## Capture modes
 
-The compiler prefers to capture a closed-over variable by immutable borrow,
+The compiler prefers to capture a value by immutable borrow,
 followed by unique immutable borrow (see below), by mutable borrow, and finally
 by move. It will pick the first choice of these that is compatible with how the
-captured variable is used inside the closure body. The compiler does not take
-surrounding code into account, such as the lifetimes of involved variables, or
+captured value is used inside the closure body. The compiler does not take
+surrounding code into account, such as the lifetimes of involved variables or fields, or
 of the closure itself.
 
-If the `move` keyword is used, then all captures are by move or, for `Copy`
-types, by copy, regardless of whether a borrow would work. The `move` keyword is
-usually used to allow the closure to outlive the captured values, such as if the
-closure is being returned or used to spawn a new thread.
+## Capture Precision
+
+The precise path that gets captured is typically the full path that is used in the closure, but there are cases where we will only capture a prefix of the path.
+
+
+### Shared prefix
+
+In the case where a path and one of the ancestor’s of that path are both captured by a closure, the ancestor path is captured with the highest capture mode among the two captures,`CaptureMode = max(AncestorCaptureMode, DescendantCaptureMode)`, using the strict weak ordering
+
+`ImmBorrow < UniqueImmBorrow < MutBorrow < ByValue`.
 
-Composite types such as structs, tuples, and enums are always captured entirely,
-not by individual fields. It may be necessary to borrow into a local variable in
-order to capture a single field:
+Note that this might need to be applied recursively.
 
 ```rust
-# use std::collections::HashSet;
-#
-struct SetVec {
-    set: HashSet<u32>,
-    vec: Vec<u32>
-}
+# fn move_value<T>(_: T){}
+let s = String::from("S");
+let t = (s, String::from("T"));
+let mut u = (t, String::from("U"));
+
+let c = || {
+    println!("{:?}", u); // u captured by ImmBorrow
+    u.1.truncate(0); // u.0 captured by MutBorrow
+    move_value(u.0.0); // u.0.0 captured by ByValue
+};
+```
 
-impl SetVec {
-    fn populate(&mut self) {
-        let vec = &mut self.vec;
-        self.set.iter().for_each(|&n| {
-            vec.push(n);
-        })
-    }
-}
+Overall the closure will capture `u` by `ByValue`.
+
+### Wild Card Patterns
+Closures only capture data that needs to be read, which means the following closures will not capture `x`
+
+```rust
+let x = 10;
+let c = || {
+    let _ = x;
+};
+
+let c = || match x {
+    _ => println!("Hello World!")
+};
```

Comment:
This seems to be wrong. Both of these examples are captured by value today -> https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4044e24fa1cffb795c2dd03239eb5fb2